### ISSUE TITLE
Custom template for remote syslog temporarily ignored when HUP is received

### ISSUE THREAD

**Issue by epontan** [balabit/syslog-ng#1423]
# syslog-ng

Docker image balabit/syslog-ng:latest

## Version of syslog-ng

```
$ docker exec -it syslog-ng syslog-ng --version
syslog-ng 3.9.1
Installer-Version: 3.9.1
Revision: 3.9.1-2
Module-Directory: /usr/lib/syslog-ng/3.9
Module-Path: /usr/lib/syslog-ng/3.9
Available-Modules: confgen,csvparser,add-contextual-data,afprog,mod-java,affile,geoip-plugin,date,pseudofile,syslogformat,json-plugin,basicfuncs,afuser,linux-kmsg-format,afstomp,afamqp,afsocket,kvformat,cef,cryptofuncs,disk-buffer,graphite,afmongodb,dbparser,sdjournal,system-source
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: on
Enable-TCP-Wrapper: on
Enable-Linux-Caps: on
```

## Platform

```
$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
ID_LIKE=archlinux
ANSI_COLOR="0;36"
HOME_URL="https://www.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
```

# Issue

## Failure

In systems where newsyslog is used to rotate log files that are written to disk by syslog-ng, it has been observed that sometimes when the HUP signal (that is sent to syslog-ng process by newsyslog) causes wrong template to be used temporarily for a remote syslog server that has been configured to use a custom template.

I suspect that there is some kind of race-condition when reloading the configuration files.

It can be reproduced quite easily using docker, with a simple syslog server in python acting as remote (catching the wrong template), a python script that generates frequent logging and shell script that sends regular HUP signals.

The custom template is defined without the program name:
```
<$PRI>$DATE $HOST : $MSG
```

Sometimes when the HUP signal is received, this format is sent to the remote server:
```
<190>Apr 10 09:45:29 0c30188d28c5 : myapp: Writing some log data 0
```

Where the expected format is to follow the custom template:
```
<190>Apr 10 09:45:29 0c30188d28c5 : Writing some log data 0
```

## Steps to reproduce

1. Start syslog-ng using docker in one terminal
```
$ docker run -it -v "$PWD/syslog-ng.conf":/etc/syslog-ng/syslog-ng.conf -v $PWD:/mnt -p 1514:1514 --name syslog-ng balabit/syslog-ng:latest
```

2. Start the simple python syslog server acting as remote on the host (with the docker local IP as argument, don't forget modify syslog-ng.conf if different) in another terminal
```
$ python pysyslog-reproduction.py 172.17.0.1
```

3. Run the reproduction shell script inside the docker container on a third terminal
```
$ docker exec -it syslog-ng bash /mnt/run-reproduction.bash
```

4. Observe the second terminal and you should get a "FOUND WRONG TEMPLATE!!" printout which means that the issue is reproduced
```
$ python pysyslog-reproduction.py 172.17.0.1
<190>Apr 10 09:44:15 617265bae7d9 : Writing some log data 0
<190>Apr 10 09:44:18 617265bae7d9 : Writing some log data 100292
<190>Apr 10 09:44:21 617265bae7d9 : Writing some log data 200462


FOUND WRONG TEMPLATE!!
<190>Apr 10 09:44:21 617265bae7d9 : myapp: Writing some log data 213854
```

## Configuration

```
$ cat syslog-ng.conf
@version: 3.9

source src {
    unix-stream("/dev/log");
    internal();
};

filter f_myapp { program("^myapp$"); };

template t_mytemplate { template("<$PRI>$DATE $HOST : $MSG\n"); template_escape(no); };

destination d_myremote {
    udp("172.17.0.1"
        port(1514)
        template(t_mytemplate)
        frac_digits(0)
        suppress(5)
    );
};

log {
    source(src);
    filter(f_myapp);
    destination(d_myremote);
};

```

## Input and output logs (if possible)

No special or strange logging from syslog-ng observed has been when the issue happens.

Scripts and configuration to reproduce can be found at [this gist](https://gist.github.com/epontan/5699c07473c6210051ba7f00ca19e778)

**Comment by epontan** [comment]
This issue seem to have been solved between version 3.10.1 and 3.11.1. I was able to bisect it down to commit 484e591e4da79a981dd3a8773e91673d0962b248 and the change make sense for the race condition observed previously.

With that and the addition of #1530 I think this issue can be closed.

**Comment by bazsi** [comment]
Hmm, I have partially reverted this commit in my
register-generators-as-plugins branch, so it would make sense to retest
with that.  #1657

**Comment by epontan** [comment]
Yes, you are right. I just re-tested on your branch and it seems commit 73c104a5fc12f9989630ec8010b5e9496a3de38e re-introduces this issue.

**Comment by bazsi** [comment]
thanks for the test. I am making a note in the pull request to check this.

-- 
Bazsi

**Comment by bazsi** [comment]
This issue has not been fixed in 3.11, that change change inadvertently closes the window of the race
and causes it to disappear.

The proper fix is in #1709

To fix it on your version, you can either cherry-pick the entire branch, or if you'd like a less intrusive patch, only use the changes on lib/template/macros.c.

Thanks a lot for the great reproduction, the issue wouldn't have been this easy to fix, as the window of the race is pretty small, and on my computer I needed roughly 3 million messages to reproduce it.

**Comment by epontan** [comment]
Really great that this issue gets a proper fix too :)
Thank you too! I really appreciate your efforts and attention put into this issue.
